File: frontend/src/api/emailBreach.ts
import { apiRequest } from "./client";

export type EmailBreachRiskLevel = "low" | "medium" | "high" | "critical";
export type EmailBreachStatus = "exposed" | "not_found" | "unknown" | "provider_error";

export interface EmailBreachDetail {
  name: string;
  domain: string;
  breach_date: string;
  data_classes: string[];
  description: string;
}

export interface EmailBreachResult {
  email: string;
  email_normalized?: string | null;
  exposed: boolean;
  status: EmailBreachStatus;
  breach_count: number;
  breaches: EmailBreachDetail[];
  risk_level: EmailBreachRiskLevel;
  recommendations: string[];
}

export interface EmailBreachFinding {
  id: string;
  title: string;
  severity: "informational" | "low" | "medium" | "high" | "critical";
  category: string;
  owasp_category: string;
  evidence: string;
  impact: string;
  recommendation: string;
  priority: number;
}

export interface EmailBreachSummary {
  total_checked: number;
  exposed_count: number;
  not_found_count: number;
  unknown_count: number;
  highest_risk: EmailBreachRiskLevel;
  top_priorities: string[];
}

export interface EmailBreachSafetyModel {
  authorized_confirmed: boolean;
  passwords_collected: boolean;
  credentials_tested: boolean;
  emails_stored: boolean;
  uploaded_files_stored: boolean;
  external_provider_used: boolean;
  note: string;
}

export interface EmailBreachCheckResponse {
  mode: string;
  provider: string;
  provider_configured: boolean;
  summary: EmailBreachSummary;
  results: EmailBreachResult[];
  findings: EmailBreachFinding[];
  safety_model: EmailBreachSafetyModel;
}

export async function checkEmailBreaches(
  emails: string[],
  authorized: boolean,
): Promise<EmailBreachCheckResponse> {
  return apiRequest<EmailBreachCheckResponse>("/email-breach/check", {
    method: "POST",
    body: { emails, authorized },
    auth: false,
  });
}

File: backend/app/schemas/email_breach.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field, field_validator


EMAIL_PATTERN = r"^[A-Za-z0-9._%+\-]+@[A-Za-z0-9.\-]+\.[A-Za-z]{2,}$"


class EmailBreachCheckRequest(BaseModel):
    emails: list[str] = Field(default_factory=list)
    authorized: bool = Field(default=False)

    @field_validator("emails")
    @classmethod
    def validate_emails_list(cls, value: list[str]) -> list[str]:
        if not value:
            raise ValueError("No valid email addresses were found.")
        if len(value) > 100:
            raise ValueError("More than 100 emails were detected. Only the first 100 will be checked.")
        return value


class EmailBreachDetail(BaseModel):
    name: str = ""
    domain: str = ""
    breach_date: str = ""
    data_classes: list[str] = Field(default_factory=list)
    description: str = ""


class EmailBreachResult(BaseModel):
    email: str
    email_normalized: str | None = None
    exposed: bool = False
    status: Literal["exposed", "not_found", "unknown", "provider_error"] = "unknown"
    breach_count: int = 0
    breaches: list[EmailBreachDetail] = Field(default_factory=list)
    risk_level: Literal["low", "medium", "high", "critical"] = "low"
    recommendations: list[str] = Field(default_factory=list)


class EmailBreachSummary(BaseModel):
    total_checked: int = 0
    exposed_count: int = 0
    not_found_count: int = 0
    unknown_count: int = 0
    highest_risk: Literal["low", "medium", "high", "critical"] = "low"
    top_priorities: list[str] = Field(default_factory=list)


class EmailBreachFinding(BaseModel):
    id: str
    title: str
    severity: Literal["informational", "low", "medium", "high", "critical"]
    category: str = "Account Exposure"
    owasp_category: str = "Identification and Authentication Failures"
    evidence: str
    impact: str
    recommendation: str
    priority: int = 1


class EmailBreachSafetyModel(BaseModel):
    authorized_confirmed: bool = True
    passwords_collected: bool = False
    credentials_tested: bool = False
    emails_stored: bool = False
    uploaded_files_stored: bool = False
    external_provider_used: bool = True
    note: str = "This tool checks email exposure using a configured provider and does not test credentials."


class EmailBreachCheckResponse(BaseModel):
    mode: str = "email_breach_check"
    provider: str = "rapidapi"
    provider_configured: bool = False
    summary: EmailBreachSummary = Field(default_factory=EmailBreachSummary)
    results: list[EmailBreachResult] = Field(default_factory=list)
    findings: list[EmailBreachFinding] = Field(default_factory=list)
    safety_model: EmailBreachSafetyModel = Field(default_factory=EmailBreachSafetyModel)
